design/pool-configuration.md

```diff
@@ -0,0 +1,552 @@
+
+##### Non overlapping filters
+
+Having an IP with no nodes advertising it is a risk. Some scenarios are:
+
+* Traffic Policy Local with pods not running on the same nodes selected by the selector
+* Node BGP peer selector with no intersection with the address selector
+
+It should be possible to detect non overlapping configurations, and raise an event when they happen. The metallb operator introduces a Validation WebHook
+that can be used to prevent those configurations to be created in first instance.
+
+### Conflicts with discovery
+
+The peer discovery feature currently being implemented in [this PR](https://github.com/metallb/metallb/pull/593) will introduce a way to
+fetch peers from the node's annotations.
+
+With this new structure, the PR will need to be adjusted to extend the `BGPPeer` entity in order to consume the node's annotations instead of hardcoded values.
+
+### Conflicts with the FRR implementation
+
+All the BGP changes will happen from a "behavioural" point of view. Where needed, the changes will reflected in both the native and the new implementation.
+
+### Not covering all the angles
+
+The risk of not covering all possible cases / permutations is still there.
+
+That aside, this proposal covers the majority of the issues that were raised by the users (multiple times), and it won't prevent to add a finer level of configuration
+if ever comes the need. Moreover, providing a straightforward way to separate concerns will make it easier to accomodate new changes, and it will be more clear where
+those new changes must be applied.
+
+## Design Details
+
+### New data model
+
+Here we describe the new proposed data model in the form of Go structures, and non comprehensive sample CRDs.
+
+#### AddressPool
+
+The `AddressPool` is the entity responsible for the IP allocation part, regardless of how (and if) the ip is going to be advertised.
+
+Hence, we keep the same structure but remove the bgpAdvertisement section and the Protocol field.
+
+```go
+type addressPool struct {
+  Name              string
+  Addresses         []string
+  AvoidBuggyIPs     bool
+  AutoAssign        *bool
+}
+```
+
+Example CRD
+
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: AddressPool
+metadata:
+  name: addresspool-sample3
+  namespace: metallb-system
+spec:
+  addresses:
+    - 2002:2:2::1-2002:2:2::100
+    - 172.20.0.100/24
+  autoAssign: false
+  avoidBuggyIPs: true
+
+```
+
+#### BGP Advertisement
+
+The `BGPAdvertisement` entity represents the intention to advertise a given IP via the BGP Protocol. It contains properties related to the type of advertisement and a reference to the address pools we want we advertise. This initial version matches the current behaviour, where an IP is advertised to all the `BGPPeer`s configured in MetalLB.
+
+```go
+type bgpAdvertisement struct {
+  Name              string
+  Pools             []string
+  PoolSelector      map[string]string
```

Comment:
Why do we have both `Pools` and `PoolSelector`? What is the difference between the two?

Reply:
My idea was to have both the simplicity of a list of names vs the flexibility of a label selector (which at this point can be expressed as `*metav1.LabelSelector`)